community/bolt/src/main/java/org/neo4j/bolt/v1/messaging/Neo4jPackV1.java

```diff
@@ -0,0 +1,609 @@
+                    int index = edgeIndexes.getOrDefault( edge.id(), NO_SUCH_ID );
+
+                    if ( node.id() == edge.startNode().id() )
+                    {
+                        pack( index );
+                    }
+                    else
+                    {
+                        pack( -index );
+                    }
+                }
+
+            }
+        }
+
+        private void writeNodesForPath( NodeValue[] nodes ) throws IOException
+        {
+            nodeIndexes.reset( nodes.length );
+            for ( NodeValue node : nodes )
+            {
+                nodeIndexes.putIfAbsent( node.id(), nodeIndexes.size() );
+            }
+
+            int size = nodeIndexes.size();
+            packListHeader( size );
+            if ( size > 0 )
+            {
+                NodeValue node = nodes[0];
+                for ( long id : nodeIndexes.keys() )
+                {
+                    int i = 1;
+                    while ( node.id() != id )
+                    {
+                        node = nodes[i++];
+                    }
+                    node.writeTo( this );
+                }
+            }
+        }
+
+        private void writeEdgesForPath( EdgeValue[] edges ) throws IOException
+        {
+            edgeIndexes.reset( edges.length );
+            for ( EdgeValue node : edges )
+            {
+                // relationship indexes are one-based
+                edgeIndexes.putIfAbsent( node.id(), edgeIndexes.size() + 1 );
+            }
+
+            int size = edgeIndexes.size();
+            packListHeader( size );
+            if ( size > 0 )
+            {
+                {
+                    EdgeValue edge = edges[0];
+                    for ( long id : edgeIndexes.keys() )
+                    {
+                        int i = 1;
+                        while ( edge.id() != id )
+                        {
+                            edge = edges[i++];
+                        }
+                        //Note that we are not doing edge.writeTo(this) here since the serialization protocol
+                        //requires these to be _unbound relationships_, thus edges without any start node nor
+                        // end node.
+                        packStructHeader( 3, UNBOUND_RELATIONSHIP );
+                        pack( edge.id() );
+                        edge.type().writeTo( this );
+                        edge.properties().writeTo( this );
+                    }
+                }
+            }
+        }
+
+        @Override
+        public void writePoint( CoordinateReferenceSystem crs, double[] coordinate ) throws IOException
+        {
+            error = new Error( Status.Request.Invalid,
+                    "Point is not yet supported as a return type in Bolt" );
+            packNull();
+        }
+
+        @Override
+        public void writeNull() throws IOException
+        {
+            packNull();
```

Comment:
Maybe we could inline all these method calls, not sure if there is a benefit of having both `pack( X Value)` and `writeX`?

Reply:
You are right, having both `#pack()` and `#write()` methods is confusing. Former come from `PackStream.Packer` which this class extends. So this class is a mix of high level methods like `#pack(AnyValue)` and low level packstream methods like `#packStructHeader()`.

I think simple inlining will make things even more confusing. This class will have to deal with `PackOutput` directly and take on some packstream functionality.

Reply:
@pontusmelke Is it correct that`Neo4jPackV1.Packer` is actually a bolt impl of `AnyValueWriter` and `PackStream.Packer` is a bolt impl of `ValueWriter`?

Then maybe the long term cleanup plan would be make PackStream.Packer impl `ValueWriter` and `Neo4jPackV1.Packer` impl `AnyValueWriter`?